cpp/src/io/json/json_tree.cu

```diff
@@ -0,0 +1,670 @@
+    // TODO if node_category is hashed, then no need for transform later. for field only, use string
+    // hash too.
+    auto field_name = cudf::string_view(d_input + node_range_begin[node_id],
+                                        node_range_end[node_id] - node_range_begin[node_id]);
+    return cudf::detail::default_hash<cudf::string_view>{}(field_name);
+  };
+  auto d_equal = [d_input          = d_input.data(),
+                  node_range_begin = d_tree.node_range_begin.data(),
+                  node_range_end   = d_tree.node_range_end.data()] __device__(auto node_id1,
+                                                                            auto node_id2) {
+    // TODO if node_category is used, then no need for transform later.
+    auto field_name1 = cudf::string_view(d_input + node_range_begin[node_id1],
+                                         node_range_end[node_id1] - node_range_begin[node_id1]);
+    auto field_name2 = cudf::string_view(d_input + node_range_begin[node_id2],
+                                         node_range_end[node_id2] - node_range_begin[node_id2]);
+    return field_name1 == field_name2;
+  };
+  auto is_field_name_node = [node_categories = d_tree.node_categories.data()] __device__(
+                              auto node_id) { return node_categories[node_id] == node_t::NC_FN; };
+  // key-value pairs: uses node_id itself as node_type. (unique node_id for a field name due to
+  // hashing)
+  auto iter = cudf::detail::make_counting_transform_iterator(
+    0, [] __device__(size_type i) { return cuco::make_pair(i, i); });
+
+  key_map.insert_if(iter,
+                    iter + num_nodes,
+                    thrust::counting_iterator<size_type>(0),  // stencil
+                    is_field_name_node,
+                    d_hasher,
+                    d_equal,
+                    stream.value());
+  auto get_hash_value =
+    [key_map = key_map.get_device_view(), d_hasher, d_equal] __device__(auto node_id) -> size_type {
+    auto it = key_map.find(node_id, d_hasher, d_equal);
+    return (it == key_map.end()) ? size_type{0} : it->second.load();
+  };
+  // convert field nodes to node indices, and other nodes to enum value.
+  rmm::device_uvector<size_type> node_type(num_nodes, stream);
+  thrust::tabulate(rmm::exec_policy(stream),
+                   node_type.begin(),
+                   node_type.end(),
+                   [node_categories = d_tree.node_categories.data(),
+                    is_field_name_node,
+                    get_hash_value] __device__(auto node_id) -> size_type {
+                     if (is_field_name_node(node_id))
+                       return static_cast<size_type>(NUM_NODE_CLASSES) + get_hash_value(node_id);
+                     else
+                       return static_cast<size_type>(node_categories[node_id]);
+                   });
+// TODO delete hash_map after node_type creation
```

Comment:
I think we would like to do that in this PR

Reply:
used IILE to scope limit the key_map.